File: src/auth/auth.service.ts
import { BadRequestException, Injectable, InternalServerErrorException, NotFoundException, UnauthorizedException } from '@nestjs/common';
import { SupabaseService } from 'src/supabase/supabase.service';
import { RegisterDto } from './dto/register.dto';
import { PrismaService } from 'src/prisma/prisma.service';
import { LoginDto } from './dto/login.dto';
import type { AuthUser } from './interfaces/auth-user.interface';
import { Prisma } from '@prisma/client';

@Injectable()
export class AuthService {
    constructor(
        private readonly supabaseService: SupabaseService,
        private readonly prisma: PrismaService,
    ) { }

    health() {
        return {
            message: 'Auth service is working',
        }
    }

    async register(dto: RegisterDto) {
        const supabase = this.supabaseService.getClient();

        const { data, error } = await supabase.auth.admin.createUser({
            email: dto.email,
            password: dto.password,
            email_confirm: true,
        });

        if (error) {
            throw new BadRequestException(error.message);
        }

        try {

            // await supabase.from('users').insert({
            //     id: data.user.id,
            //     email: dto.email,
            //     full_name: dto.fullName,
            //     phone_number: dto.phoneNumber,
            //     market_location: dto.marketLocation,
            // });

            await this.prisma.users.create({
                data: {
                    id: data.user.id,
                    email: dto.email,
                    full_name: dto.fullName,
                    phone_number: dto.phoneNumber,
                    market_location: dto.marketLocation,
                }
            })

            return {
                message: 'User registered successfully',
                user: data.user,
            };
        } catch (err) {
            console.error('Prisma Error:', err);
            await supabase.auth.admin.deleteUser(data.user.id)
            if (
                err instanceof Prisma.PrismaClientKnownRequestError &&
                err.code === 'P2002'
            ) {
                const field = (err.meta?.target as string[])?.[0];

                throw new BadRequestException(
                    `${field} already exists`,
                );
            }
            throw new InternalServerErrorException(
                'Failed to create user profile',
            );

        }
    }


    async login(dto: LoginDto) {
        const supabase = this.supabaseService.getClient();

        const { data, error } =
            await supabase.auth.signInWithPassword({
                email: dto.email,
                password: dto.password,
            });

        if (error || !data.session) {
            throw new UnauthorizedException(
                'Invalid email or password',
            );
        }

        const profile = await this.prisma.users.findUnique({
            where: {
                id: data.user.id,
            },
        });

        if (!profile) {
            throw new NotFoundException(
                'User profile not found',
            );
        }

        return {
            message: 'Login successful',
            data: {
                accessToken: data.session.access_token,
                refreshToken: data.session.refresh_token,
                expiresIn: data.session.expires_in,
                user: profile,
            },
        };
    }

    async googleLogin(redirectTo: string) {
        const supabase = this.supabaseService.getClient();

        const { data, error } = await supabase.auth.signInWithOAuth({
            provider: 'google',
            options: {
                redirectTo: redirectTo,
            },
        });

        if (error) {
            throw new BadRequestException(error.message);
        }

        return {
            url: data.url,
        };
    }


    async completeGoogle(
        user: AuthUser,
    ) {

        let profile =
            await this.prisma.users.findUnique({
                where: {
                    id: user.id,
                },
            });

        if (!profile) {

            profile =
                await this.prisma.users.create({

                    data: {
                        id: user.id,
                        email: user.email,
                        phone_number: ""
                    },

                });

        }

        return {
            message: 'Google login successful',
            user: profile,
        };

    }
}

File: src/services/auth_service.py
from fastapi import HTTPException
from postgrest.exceptions import APIError

from src.schemas.auth import LoginRequest, RegisterRequest, SendOtpRequest, VerifyOtpRequest
from src.services.supabase_client import get_auth_client, get_supabase_client


def register(body: RegisterRequest) -> dict:
    client = get_supabase_client()

    result = client.auth.admin.create_user(
        {"email": body.email, "password": body.password, "email_confirm": True}
    )
    user = result.user
    if not user:
        raise HTTPException(status_code=400, detail="Registration failed")

    try:
        client.table("users").insert(
            {
                "id": user.id,
                "email": body.email,
                "full_name": body.full_name,
                "phone_number": body.phone_number,
                "market_location": body.market_location,
            }
        ).execute()
    except APIError as e:
        client.auth.admin.delete_user(user.id)  # roll back the auth user so retry is possible
        if e.code == "23505":  # Postgres unique_violation
            raise HTTPException(status_code=400, detail="Email or phone number already exists")
        raise HTTPException(status_code=500, detail="Failed to create user profile")

    return {"message": "User registered successfully", "user": user.model_dump()}


def login(body: LoginRequest) -> dict:
    # Sign-in on a throwaway client — sign_in_with_password persists a session on
    # whatever client it's called on, and we must never let that touch the shared
    # service-role client used for every other request.
    auth_client = get_auth_client()

    try:
        result = auth_client.auth.sign_in_with_password({"email": body.email, "password": body.password})
    except Exception:
        raise HTTPException(status_code=401, detail="Invalid email or password")

    if not result.session:
        raise HTTPException(status_code=401, detail="Invalid email or password")

    # Profile lookup goes through the cached service-role client, so it always
    # runs with admin privileges regardless of the throwaway client's session.
    client = get_supabase_client()
    profile = client.table("users").select("*").eq("id", result.user.id).execute()
    if not profile.data:
        raise HTTPException(status_code=404, detail="User profile not found")

    return {
        "message": "Login successful",
        "data": {
            "accessToken": result.session.access_token,
            "refreshToken": result.session.refresh_token,
            "expiresIn": result.session.expires_in,
            "user": profile.data[0],
        },
    }


def send_otp(body: SendOtpRequest) -> dict:
    auth_client = get_auth_client()
    try:
        auth_client.auth.sign_in_with_otp({"phone": body.phone_number})
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {"message": "OTP sent successfully"}


def verify_otp(body: VerifyOtpRequest) -> dict:
    # Same reasoning as login(): verify_otp establishes a session, so it must run
    # on a throwaway client, never the shared service-role one.
    auth_client = get_auth_client()

    try:
        result = auth_client.auth.verify_otp(
            {"phone": body.phone_number, "token": body.token, "type": "sms"}
        )
    except Exception as e:
        raise HTTPException(status_code=401, detail=str(e))

    if not result.user:
        raise HTTPException(status_code=401, detail="User not found")

    client = get_supabase_client()
    _ensure_profile_exists(client, result.user.id, result.user.email)

    return {
        "accessToken": result.session.access_token if result.session else None,
        "refreshToken": result.session.refresh_token if result.session else None,
        "user": result.user.model_dump(),
    }


def _ensure_profile_exists(client, user_id: str, email: str | None) -> None:
    """First-time phone login creates a bare users row, same as the Node backend does."""
    existing = client.table("users").select("id").eq("id", user_id).execute()
    if not existing.data:
        client.table("users").insert({"id": user_id, "email": email, "phone_number": ""}).execute()
